scylla-cdc-base/src/main/java/com/scylladb/cdc/model/worker/TaskAction.java

```diff
@@ -41,97 +64,103 @@ protected ReadNewWindowTaskAction(WorkerConfiguration workerConfiguration, Task
         }
     }
 
     private static class ReadChangeTaskAction extends TaskAction {
-        private final WorkerConfiguration workerConfiguration;
-        private final Task task;
         private final Reader reader;
         private final int tryAttempt;
-
-        public ReadChangeTaskAction(WorkerConfiguration workerConfiguration, Task task, Reader reader, int tryAttempt) {
-            this.workerConfiguration = Preconditions.checkNotNull(workerConfiguration);
-            this.task = Preconditions.checkNotNull(task);
+        private final TaskState newState;
+        
+        public ReadChangeTaskAction(WorkerConfiguration workerConfiguration, Task task, Reader reader, int tryAttempt, TaskState newState) {
+            super(workerConfiguration, task);
             this.reader = Preconditions.checkNotNull(reader);
             Preconditions.checkArgument(tryAttempt >= 0);
             this.tryAttempt = tryAttempt;
+            this.newState = newState;
         }
         
+        private CompletableFuture<TaskAction> onException(Throwable ex) {
+            // Exception occured while reading the window, we will have to restart
+            // ReadNewWindowTaskAction - read a window from state defined in task.
+            long backoffTime = workerConfiguration.workerRetryBackoff.getRetryBackoffTimeMs(tryAttempt);
+            logger.atSevere().withCause(ex).log("Error while reading a CDC change. Task: %s. " +
+                    "Task state: %s. Will retry after backoff (%d ms).", task.id, task.state, backoffTime);
+            return delay(backoffTime)
+                    .thenApply(t -> new ReadNewWindowTaskAction(workerConfiguration, task, tryAttempt + 1));
+        }
+
         @Override
         public CompletableFuture<TaskAction> run() {
-            workerConfiguration.transport.setState(task.id, task.state);
-
-            CompletableFuture<TaskAction> taskActionFuture = reader.nextChange().
-                    thenApply(change -> new ConsumeChangeTaskAction(workerConfiguration, task, reader, change, tryAttempt));
-            return FutureUtils.thenComposeExceptionally(taskActionFuture, ex -> {
-                // Exception occured while reading the window, we will have to restart
-                // ReadNewWindowTaskAction - read a window from state defined in task.
-                long backoffTime = workerConfiguration.workerRetryBackoff.getRetryBackoffTimeMs(tryAttempt);
-                logger.atSevere().withCause(ex).log("Error while reading a CDC change. Task: %s. " +
-                        "Task state: %s. Will retry after backoff (%d ms).", task.id, task.state, backoffTime);
-                return workerConfiguration.delayedFutureService.delayedFuture(backoffTime)
-                        .thenApply(t -> new ReadNewWindowTaskAction(workerConfiguration, task, tryAttempt + 1));
-            });
+            if (newState != null) {
+                workerConfiguration.transport.setState(task.id, newState);
```

Comment:
This `setState(task.id, newState)` instead of (previously) `setState(task.id, task.state)` also introduces a subtle bug that was not present before:

Consider a window with 2 changes. Before this commit, when reading such a window, the state in transport would be updated 3 times:
1. `TaskState(windowStart, windowEnd, lastConsumed=Optional.empty())`
2. `TaskState(windowStart, windowEnd, lastConsumed=Optional.of(change1.getId()))`
3. `TaskState(windowStart, windowEnd, lastConsumed=Optional.of(change2.getId()))`  

After this commit, the state is updated as follows:
1. `TaskState(windowStart, windowEnd, lastConsumed=Optional.empty())`
2. `TaskState(windowStart, windowEnd, lastConsumed=Optional.empty())`
3. `TaskState(windowStart, windowEnd, lastConsumed=Optional.of(change1.getId()))`

This can cause unnecessary duplication of read CDC changes after `Worker` restart.

This is an example of why I was opposing `Consumer` returning `TaskState` - it could be error-prone and hard to reason about.

Reply:
Uhm, the "normal" (`TaskAndRawChangeConsumer` wrapping) consume just returns the task.state as newState. I.e. newState in this patch set is _always_ task.state. How does this change the reporting (`setState`) calls?

Reply:
Ah, now I see it. That is borked. But that is not an issue of the state being returned, it is an issue with which task we are looking at.

Reply:
Fixed.